elevation_layer/src/elevation_layer.cpp

```diff
@@ -0,0 +1,242 @@
+
+    ElevationLayer::~ElevationLayer()
+    {
+        if (dsrv_)
+            delete dsrv_;
+    }
+
+    void ElevationLayer::updateBounds(double robot_x, double robot_y, double robot_yaw, double* min_x, double* min_y,
+                              double* max_x, double* max_y)
+    {
+        std::lock_guard<std::mutex> lock(elevation_map_mutex_);
+        if (rolling_window_)
+            updateOrigin(robot_x - getSizeInMetersX() / 2, robot_y - getSizeInMetersY() / 2);
+        if (!(enabled_ && elevation_map_received_))
+            return;
+        useExtraBounds(min_x, min_y, max_x, max_y);
+
+        for (grid_map::GridMapIterator iterator(elevation_map_); !iterator.isPastEnd(); ++iterator) {
+            const grid_map::Index gridmap_index(*iterator);
+            grid_map::Position vertexPositionXY;
+            elevation_map_.getPosition(gridmap_index, vertexPositionXY);
+            double px = vertexPositionXY.x();
+            double py = vertexPositionXY.y();
+
+            touch(px, py, min_x, min_y, max_x, max_y);
+        }
+        updateFootprint(robot_x, robot_y, robot_yaw, min_x, min_y, max_x, max_y);
+    }
+
+    void ElevationLayer::updateFootprint(double robot_x, double robot_y, double robot_yaw, double* min_x, double* min_y,
+                                        double* max_x, double* max_y)
+    {
+        if (!footprint_clearing_enabled_) return;
+        costmap_2d::transformFootprint(robot_x, robot_y, robot_yaw, getFootprint(), transformed_footprint_);
+
+        for (unsigned int i = 0; i < transformed_footprint_.size(); i++)
+        {
+            touch(transformed_footprint_[i].x, transformed_footprint_[i].y, min_x, min_y, max_x, max_y);
+        }
+    }
+
+    void ElevationLayer::updateCosts(costmap_2d::Costmap2D& master_grid, int min_i, int min_j, int max_i, int max_j)
+    {
+        std::lock_guard<std::mutex> lock(elevation_map_mutex_);
+        if (!enabled_ || !elevation_map_received_)
+            return;
+        const bool has_edges_layer = elevation_map_.exists(edges_layer_name_);
+        const grid_map::Matrix& elevation_data = elevation_map_[elevation_layer_name_];
+        for (grid_map::GridMapIterator iterator(elevation_map_); !iterator.isPastEnd(); ++iterator) {
+            const grid_map::Index gridmap_index(*iterator);
+            grid_map::Position vertexPositionXY;
+            elevation_map_.getPosition(gridmap_index, vertexPositionXY);
+            double px = vertexPositionXY.x();
+            double py = vertexPositionXY.y();
+            // now we need to compute the map coordinates for the observation
+            unsigned int mx, my;
+            if (!worldToMap(px, py, mx, my))    // if point outside of local costmap, ignore
+            {
+                continue;
+            }
+            if ( elevation_data(gridmap_index(0), gridmap_index(1)) > height_treshold_ )  // If point too high, it could be an obstacle
+            {
+                if (!has_edges_layer) {
```

Comment:
You could check this before even going into the for loop, since this is a property of the elevation_map. Also use a throttled message.